alf/algorithms/planning_algorithm.py

```diff
@@ -269,3 +282,126 @@ def _calc_cost_for_action_sequence(self, time_step: TimeStep, state,
 
     def after_update(self, training_info):
         pass
+
+
+@gin.configurable
+class CEMPlanAlgorithm(RandomShootingAlgorithm):
+    """CEM-based planning method.
+
+    This method uses a Cross-Entropy Method (CEM) to optimize an action
+    trajectory by minimizing a given cost function. The optimized action
+    trajectory is termed as a 'plan' which can be used by other components
+    such as a MPC-based controller. This has been used by some MBRL works
+    such as `Deep Reinforcement Learning in a Handful of Trials using
+    Probabilistic Dynamics Models <https://arxiv.org/abs/1805.12114>`_
+
+    To speedup, when possible, we have used the plan obtained at the previous
+    time step to initialize the the mean of the plan distribution at the current
+    time step, after proper shifting and padding. This, however, makes the
+    current implementation no longer thread-safe.
```

Comment:
Perhaps we can use state to handle this to make it thread-safe

Reply:
Good point. Will try to switch to state.

Reply:
Now switched to use state for passing the previous plan.